File: String/Easy/383. Ransom Note.go
package main

import (
	"fmt"
	"strings"
)

func main() {

	ransomNote, magazine := "aa", "ab"
	result := canConstruct(ransomNote, magazine)
	fmt.Println(result)
}

func canConstruct(ransomNote string, magazine string) bool {
	ransom := strings.Split(ransomNote, "")
	mag := strings.Split(magazine, "")

	ansMap := make(map[string]int)
	for _, val := range mag {
		ansMap[val]++
	}

	for _, val := range ransom {
		count, ok := ansMap[val]
		if !ok || count < 1 {
			return false
		}
		ansMap[val]--
	}
	return true
}

File: Daily Challenges/383. Ransom Note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        dict = {}
        
        for char in magazine:
            if char not in dict.keys():
                dict[char] = 1
            else:
                dict[char] += 1
        
        for i in range(len(ransomNote)):
            if ransomNote[i] not in dict.keys():
                return False
            elif dict[ransomNote[i]] < 1:
                return False
            else:
                dict[ransomNote[i]] -= 1
        
        return True

obj = Solution()
ransomNote, magazine = "aa", "aab"
ans = obj.canConstruct(ransomNote, magazine)
print(ans)